File: downloader/lib/captions.py
"""Instagram caption generator, hard-locked to Ahmedabad interior/real-estate niche."""

import os

from openai import OpenAI

DEFAULT_MODEL = "gpt-4o-mini"


def _client() -> OpenAI:
    key = (os.environ.get("OPENAI_API_KEY") or "").strip()
    if not key:
        raise RuntimeError("OPENAI_API_KEY is not set — check .env")
    return OpenAI(api_key=key)


def _city() -> str:
    return (os.environ.get("NICHE_CITY") or "Ahmedabad").strip()


def _topics() -> str:
    return (os.environ.get("NICHE_TOPICS") or "real estate, interior design").strip()


def _system_prompt() -> str:
    city = _city()
    topics = _topics()
    return f"""You are an Instagram copywriter for an {city}-based {topics} brand.

Every caption you write promotes {topics} services in {city}, India. The target audience is homeowners, home-buyers, and property investors in {city} and nearby Gujarat.

Output format (return ONLY the caption — no quotes, no markdown, no explanations):
- Line 1: a hook (question, bold statement, or relatable home-buying / decor pain point). 8–14 words.
- Lines 2–3: 1–2 short aesthetic lines that describe the visual + tie it to living in {city} (neighbourhoods like Bodakdev, SG Highway, Prahlad Nagar, Satellite, Thaltej, South Bopal, Shela, Gota, Vastrapur are fine to reference when relevant — but only occasionally, not every caption).
- Line 4: a soft call-to-action — DM for a free consultation / site visit / quote, save this, or comment a room name.
- Blank line, then 18–25 hashtags on a single line. Mix:
    • 4–6 high-volume: #interiordesign #realestate #homedecor #luxuryhomes #modernhome #dreamhome
    • 6–8 {city}/Gujarat-specific: #ahmedabad #ahmedabaddiaries #ahmedabadrealestate #ahmedabadinteriors #gujarat #amdavadi #ahmedabadhomes
    • 6–8 niche: match the video (kitchen, bedroom, pooja room, living room, 2bhk, 3bhk, villa, flat, bungalow, wardrobe, modular, tv unit, etc.)

Hard rules:
- Always real estate OR interior design — never anything else.
- Always {city} / Gujarat framing.
- English + light Hinglish allowed ("ghar", "apna home"). Keep it natural, not cringe.
- No AI/automation/tool mentions. No "meet our model" framing.
- Total length before hashtags: 150–280 characters.
- Max 2–3 emojis, only if they fit (🏡 🛋️ 🪴 ✨ are safe).
"""


def _user_prompt(keyword: str, title: str | None) -> str:
    title_part = f"Video context / title: {title.strip()}\n" if title and title.strip() else ""
    return (
        f"Search keyword that pulled this video: {keyword}\n"
        f"{title_part}"
        f"Write the Instagram caption."
    )


def generate_caption(keyword: str, title: str | None = None) -> tuple[str, str]:
    """Return (caption_text, model_name)."""
    model = (os.environ.get("OPENAI_CAPTION_MODEL") or DEFAULT_MODEL).strip()
    client = _client()

    resp = client.chat.completions.create(
        model=model,
        temperature=0.8,
        max_tokens=500,
        messages=[
            {"role": "system", "content": _system_prompt()},
            {"role": "user", "content": _user_prompt(keyword, title)},
        ],
    )
    text = (resp.choices[0].message.content or "").strip()
    if not text:
        raise RuntimeError("OpenAI returned empty caption")
    # Normalise whitespace
    text = text.replace("\r\n", "\n").strip()
    return text, model

File: lib/captions.ts
import "server-only";

import OpenAI from "openai";

const DEFAULT_MODEL = "gpt-4o-mini";

function client() {
  const apiKey = (process.env.OPENAI_API_KEY || "").trim();
  if (!apiKey) throw new Error("OPENAI_API_KEY is not set");
  return new OpenAI({ apiKey });
}

function city() {
  return (process.env.NICHE_CITY || "Ahmedabad").trim();
}

function topics() {
  return (process.env.NICHE_TOPICS || "real estate, interior design").trim();
}

function systemPrompt() {
  const c = city();
  const t = topics();
  return `You are an Instagram copywriter for an ${c}-based ${t} brand (@style_o_studio).

Every caption must:
- Promote ${t} services in ${c}, India ONLY. Audience: homeowners, home-buyers, and property investors in ${c} and nearby Gujarat.
- End with a CTA that explicitly asks the reader to DM if they want ${t} like what's shown. Example phrasings: "DM us to create this home", "Want a home like this? DM us in ${c}", "Book a free consultation — DM us".

Output format (return ONLY the caption — no quotes, no markdown, no preamble):

[Hook line] — a question, bold claim, or relatable home-buying / decor pain point. 8–14 words.

[2–3 aesthetic body lines] describing what's in the video + tying it to ${c} living. Reference a neighbourhood when it fits naturally (pick one only when relevant, not every time):
Bodakdev · SG Highway · Prahlad Nagar · Satellite · Thaltej · South Bopal · Shela · Gota · Vastrapur · Sindhu Bhavan Road · Science City · Chandkheda · Naranpura · Paldi.

[Key details line] — 2-4 short factual call-outs separated by · (e.g. "Matte-finish modular · Quartz countertop · Under-cabinet LEDs · ${c} built 2024"). Make these specific to the video — don't repeat generic words.

[CTA line] — one sentence. Must explicitly invite a DM to create this type of home. Examples:
"DM us to create a home like this in ${c} 🏡"
"Want this for your home? DM @style_o_studio — based in ${c}."

[blank line]

[hashtags] — 22–28 hashtags on a single line, no commas. Mix:
  • 6–8 high-volume: #interiordesign #realestate #homedecor #luxuryhomes #modernhome #dreamhome #homeinspo #interiorinspo
  • 8–10 ${c}/Gujarat-specific: #ahmedabad #ahmedabaddiaries #ahmedabadrealestate #ahmedabadinteriors #ahmedabadhomes #gujarat #amdavadi #ahmedabadinteriordesigner #gujaratinteriors #ahmedabadlife
  • 6–10 niche-to-video (pick words that match the room/feature shown): #modularkitchen #poojaroom #bedroomdecor #livingroom #tvunit #wardrobe #2bhk #3bhk #villa #bungalow #jaliwork #fluteddesign #archdesign #walltexture etc.

Hard rules:
- Caption text before hashtags: 220–360 characters (count spaces).
- Exactly ONE blank line between the caption body and the hashtags.
- English + light Hinglish OK ("ghar", "apna home").
- Max 2–3 emojis total. Safe: 🏡 🛋️ ✨ 🪴 💫.
- Never mention AI, automation, tools, or this process. Never say "Meet our model".
- Always push ${c} / Gujarat framing — no other cities.`;
}

export type CaptionResult = { text: string; model: string };

export async function generateCaption(input: {
  keyword: string;
  title?: string | null;
}): Promise<CaptionResult> {
  const model = (process.env.OPENAI_CAPTION_MODEL || DEFAULT_MODEL).trim();
  const c = client();

  const userLines = [
    `Search keyword that pulled this video: ${input.keyword}`,
  ];
  if (input.title && input.title.trim()) userLines.push(`Video context / title: ${input.title.trim()}`);
  userLines.push("Write the Instagram caption following every rule in the system prompt.");

  const res = await c.chat.completions.create({
    model,
    temperature: 0.85,
    max_tokens: 800,
    messages: [
      { role: "system", content: systemPrompt() },
      { role: "user", content: userLines.join("\n") },
    ],
  });

  const text = (res.choices[0]?.message?.content || "").trim();
  if (!text) throw new Error("OpenAI returned empty caption");
  return { text, model };
}
